### DISCUSSION TITLE
For us design luvvers that do get itchy...

### DISCUSSION THREAD

**Discussion by bijlevel** [RomanKubiak/ctrlr#623] [Ideas]
https://github.com/RomanKubiak/ctrlr/discussions/596#discussioncomment-8152942

@Ted-juh Leef je nog?😋

I have studied your post (see link above in this post) about the custom rotary slider, because I like the design and I'm at the moment busy with a panel design. Am I right in my understanding that the outer red ring is a L&F v4 slider and the rest a "painted" rotary slider. So a combination? Maybe you could post a panel-example with this slider? 

This is wat I'm aiming at (very roughly, colours have to change...), a L&F v4 rotary combined with some button image:

<img width="140" alt="custom rotary" src="https://github.com/RomanKubiak/ctrlr/assets/11859363/6042c623-3363-4be4-a6ad-6c4d3c4e999b">

**Comment by damiensellier** [comment]
I've never seen a knob like this :)
You have 4 circles in there, that's a lot. I'm not sure what you want to do. Is it like a dual-concentric/stacked knob? 
If not you need to use the slider lookandfeelmethod drawRotarySlider() to draw your different elements.

https://docs.juce.com/master/structSlider_1_1LookAndFeelMethods.html#ae63a9d8adce084cd5dbe02b960c73a9a
https://docs.juce.com/master/tutorial_look_and_feel_customisation.html

Some custom exemples for the sliders are available in the JUCE demorunner in the lookandfeel category.

**Reply by bijlevel** [reply]
I'm no sure if you mean the knob of @Ted-juh or 'my not yet functional' knob picture. If mine, then I just wanted to get rid of the flatness of the standard Juce rotary slider by filling up the empty space in the middle with some 3D knob representation. But looking at the code of Ted-juh it is probably possible to design it completely within Juce/Lua without using the standard slider.

**Comment by Ted-juh** [comment]
Sorry for the delay. Got home late last night and crashed right away.

I decided against posting the whole panel because it would make it unnecessarily complicated. But here is the L&F for "that" button:

`
slidLF1 = {

	drawRotarySlider = function(g,x,y,width,height,sliderPosProportional,rotaryStartAngle,rotaryEndAngle,slider)

	------------------------------------------------------------------------------------------------------------

	centre = width * 0.5					-- centre of component, given that width and height are the same
	innerSize = width * 0.35				-- inner ring of dial to hold the image for the knob

	------------------------------------------------------------------------------------------------------------

	diam4		= width * 0.65
	rad4		        = diam4 * 0.5
	rotEll		= 0
	frRad		= 0
	toRad		= math.pi * 2
	pathStr4	= width * 0.04

	grad1x 		= width * 0.6
	grad1y		= width * 0.275
	grad2x		= width * 0.5
	grad2y		= 0.67

    slidMin = rotaryStartAngle
    slidMax = rotaryStartAngle + sliderPosProportional * (rotaryEndAngle - rotaryStartAngle)

	arcP = Path()
	g:setColour(Colour(0xFFFF0000))
	arcP:addCentredArc(centre , centre, rad4, rad4, rotEll, slidMin, slidMax, true )
	g:strokePath(arcP, PathStrokeType(pathStr4), AffineTransform())

	------------------------------------------------------------------------------------------------------------

	-- (3a) Knob, Inner Black Circle, because of transparancy, need to fill a black bg.

	diam3 	= width * 0.55
	rad3 	= diam3 * 0.5
	iRC3x	= centre - rad3

	grad1x	= width * 0.5
	grad1y	= 0
	grad2x 	= width * 0.5
	grad2y	= width * 0.4

	circHolder = Path()
	g:setColour(Colour(0xC8000000))
	circHolder:addEllipse(iRC3x,iRC3x,diam3,diam3)
	g:fillPath(circHolder, AffineTransform())

	------------------------------------------------------------------------------------------------------------

	-- (3b) Knob, Inner Red Circle

	diam3 	= width * 0.55
	rad3 	= diam3 * 0.5
	iRC3x	= centre - rad3

	grad1x	= width * 0.6
	grad1y	= 0
	grad2x 	= width * 0.5
	grad2y	= width * 0.53

	circHolder = Path()
	gradient1 = ColourGradient(Colour(0x99FFFFFF), grad1x, grad1y, Colour(0x609E282B) , grad2x, grad2y, true)
	g:setGradientFill(gradient1)
	circHolder:addEllipse(iRC3x,iRC3x,diam3,diam3)
	g:fillPath(circHolder, AffineTransform())

	------------------------------------------------------------------------------------------------------------

	-- (4) Red Outer Ring Around Inner Knob
	-- this uses addCentredArc, drawEllipse can also be used and is even more convenient
	-- g:drawEllipse(rx,ry,rw,rh,1.0) 1.0 = thickness of the outer circle

	diam4		= width * 0.45
	rad4		= diam4 * 0.5
	rotEll		= 0
	frRad		= 0
	toRad		= math.pi * 2
	pathStr4	= width * 0.03

	grad1x 		= width * 0.6
	grad1y		= width * 0.275
	grad2x		= width * 0.5
	grad2y		= 0.67

	arcP = Path()
	gradient3 = ColourGradient(Colour(0xFFFFFFFF), grad1x, grad1y, Colour(0xFF7D0000) , grad2x, grad2y, true)
	g:setGradientFill(gradient3)
	arcP:addCentredArc(centre , centre, rad4, rad4, rotEll, frRad, toRad, true )
	g:strokePath(arcP, PathStrokeType(pathStr4), AffineTransform())

	------------------------------------------------------------------------------------------------------------

	-- (5) Outer Ring, Silver Shiny
	-- see 4 for drawEllipse comment.

	diam5 		= width * 0.55
	rad5 		= diam3 * 0.5
	iRC5x		= centre - rad3
	rotEll		= 0
	frRad		= 0
	toRad		= math.pi * 2
	pathStr5	= width * 0.06

	grad1x		= width * 0.6
	grad1y		= width * 0.1
	grad2x		= width * 0.5
	grad2y		= width * 0.67

	gradient2 = ColourGradient(Colour(0xFFFFFFFF), grad1x, grad1y, Colour(0xFF2b2b2b) , grad2x, grad2y, true)
	g:setGradientFill(gradient2)
	arcP = Path()
	arcP:addCentredArc(centre , centre, rad5, rad5, rotEll, frRad, toRad, true )
	g:strokePath(arcP, PathStrokeType(pathStr5), AffineTransform())

	------------------------------------------------------------------------------------------------------------

	-- (6) sliderpointer.. not the same way as in the juce tutorials but does the same.

	angle = rotaryStartAngle + sliderPosProportional * (rotaryEndAngle - rotaryStartAngle)

	g:setColour(Colour(0xFFFF0000))
	pointert	= Path()
	pointert:addRectangle(centre , width * 0.2 , width * 0.03, width * 0.06)
	rotation = AffineTransform.rotation(angle ,centre,centre)
	g:fillPath(pointert,rotation)

	end,
}
`
The complicating factor in this L&F is the gradient and transparency to get that "3D Metallic" effect but once you get the hang of it, it's quite easy. Remember that if you use the graphics class in Juce, you paint from bottom to top. 

In this case, the outer "red" ring is a visual aid because the "pointer" got very small depending on the button's size. It fills the value from 0  to where the pointer is. With a little help from drawText, you can even show the Value of the slider in the middle of the button instead of TextBoxBelow. Either way, you can go crazy with those designs. And the beauty of it all is that it's just paint routines and you can resize the button freely.

@damiensellier: A button never seen before makes it a unique button ;-).

**Reply by bijlevel** [reply]
@Ted-juh: pet-juh af!

Thanks a lot for the code! I will try to implement it, will take probably some time, and let you know about my experiences and "designs".

**Reply by Ted-juh** [reply]
I'll post some pictures later tonight of a few designs I made. Mr dnaldoog also made a cool design. But those experiments were almost 2 years ago(?) so I need to dig deep in my test-case folders to find them.

**Comment by Ted-juh** [comment]
A few designs. Some were solely made to figure out movement in combination with the Graphics Class. Others were made as an understudy to use in a panel later. 
![Pitch1](https://github.com/RomanKubiak/ctrlr/assets/71118185/f94322df-6709-4bb9-8f57-a98f9e14d3b8)
![LinearSlider1](https://github.com/RomanKubiak/ctrlr/assets/71118185/7d1b678c-a7c3-4607-a6fa-41192db90aea)
![OvalMcOvalFace](https://github.com/RomanKubiak/ctrlr/assets/71118185/9e6d0fcd-755d-42f3-a3e4-3e75bfbf5ce6)
![WaveShapeMovement](https://github.com/RomanKubiak/ctrlr/assets/71118185/29da6324-b31d-47e9-84d4-6ef388e2a6c3)

And a design from Dnaldoog: 
![dnaldoog](https://github.com/RomanKubiak/ctrlr/assets/71118185/b1029b23-4ff6-4464-8c41-fb1673c7bb11)

**Reply by bijlevel** [reply]
Wom man, never knew that this was possible in CTRLR! Very inspiring! Also compliments to Dnaldoog's design, comes close to what I want to achieve.

**Reply by dnaldoog** [reply]
That pitch dial is pretty cool!

**Comment by dnaldoog** [comment]
Hi @bijlevel Here are some panels Tedjuh and I were working on, including:

 
![image](https://github.com/RomanKubiak/ctrlr/assets/76831683/c3ecc71b-cbc2-437f-8cb3-530f0a128422)


Enjoy.

[My Look and feel experiments_tedjuh dnaldoog.zip](https://github.com/RomanKubiak/ctrlr/files/14731252/My.Look.and.feel.experiments_tedjuh.dnaldoog.zip)

**Reply by bijlevel** [reply]
Thanks a lot! Will be a great source of inspiration, I'm sure! One of the advantages over Knobman is of course scalability (while keeping the resolution) and probably flexibility (and even more advantages which I will encounter).

**Reply by dnaldoog** [reply]
That’s true. I try to use bitmap based graphics as little as possible these days.

**Reply by bijlevel** [reply]
I did look at the various demo panels and the one called 'My look and feel experiments' looks very promising, beause one can set the various graphic aspects of a component by changing the value of the sliders 'aaa' to 'hhh'. Took some time to figure out how it works exactly, but I gradually got the hang of it 😋.  A question though: how can it be adapted to a horizontal or vertical slider?

**Reply by Ted-juh** [reply]
Well, the answer is in Dnaldoog's reply. Set the slider style to linearBarVertical or linearBarHorizontal. Or any of the other ones that can be set as slider style from the Juce slider class. 

Not all can be set in Ctrlr. For example the threeValueSlider but if you apply it with Lua even that one shows up. But if I remember correctly not all sliders are implemented in the L&F for Ctrlr. Therefore the threeValueSlider will show up but only in the L&F that's applied to the whole panel.

If you mean horizontal or vertical as the way you have to drag it, that also depends on the slider style. But Dnaldoog found a way to drag that's both horizontal and vertical. I think you can call it multi-directional. Verry awesome, and to be honest, a more intuitive way to control a slider. Especially when a panel is close to the edge of the screen and you can't drag the mouse any further.

I can post that later if John doesn't mind. Or he can post it himself because, again, I have to weed through a lot of panels to find the right file.

**Comment by dnaldoog** [comment]
Here's another less impressive but very useful piece of code. You can create vertical linear bars using code like:

`panel:getSlider("lfo rate"):getOwnedSlider():setSliderStyle(Slider.LinearBarVertical)`

on panel load. The trick here is to set `Value background colour` to `00000000`

I don't think this can be set in the editor as far as I can see. Linear Bar always goes left to right.

----
![linearbarvertical](https://github.com/RomanKubiak/ctrlr/assets/76831683/e83bdc63-5fab-48f1-8126-6eee238118ec)

[Set Linear Bar Vertical_1.zip](https://github.com/RomanKubiak/ctrlr/files/14731359/Set.Linear.Bar.Vertical_1.zip)

**Comment by dobo365** [comment]
Very interesting...Especially the wave selector and the linear vertical bars.
For all the buttons I'm just using jknobman but indeed it is different...
For the vertical bars, I drew them in a custom component but they are static. I may replace that using your trick to have them dynamic.

**Reply by Ted-juh** [reply]
The wave selector is a custom component. If I remember correctly it uses a timer to get the waveshape to loop because it's just 2 or 2,5 squares, blocks or waves. Figured it out with the help of Dnaldoog.

**Comment by dnaldoog** [comment]
> Very interesting...Especially the wave selector and the linear vertical bars. For all the buttons I'm just using jknobman but indeed it is different... For the vertical bars, I drew them in a custom component but they are static. I may replace that using your trick to have them dynamic.

I use it often because they do look nice, design wise. But I don’t think I could’ve discovered that trick - it must’ve been Tedjuh who discovered it when we were working on ‘look and feel’ a while back. Just to give credit where credit is due!

**Reply by Ted-juh** [reply]
Coming up with, sometimes crazy, ideas isn't that hard. But to put them to code is a different thing. It was a successful partnership. So, credits where credits are due.
